### DISCUSSION TITLE
Improving how parts are added to designs

### DISCUSSION THREAD

**Discussion by eriese** [freesewing/freesewing#2538] [Forum]
@joostdecock suggested in the discussion on #2497 that we start thinking about an improved approach to configuring parts on a design. I think it's related to #1107, but a broader question, so I'm opening a new topic.

Here's my proposal:

We make a class in core to manage creating and extending design configurations. This allows us to stay away from making designers navigate the dangers of class inheritance but allows a little more _magic_. The biggest change to the design configuration itself is that instead of `dependencies`, `hide`, and `inject`, we now use part configuration objects that have all that info and more, which are exported from the individual part files. They are stored in a `parts` object in the configuration, keyed with their names, and they look like this:
``` javascript
/// back.js
export default {
  name: 'back',
  inject: 'front',
  dependencies: [],
  hidden: false,
  measurements: [],
  optionalMeasurements: [],
  options: {},
  optionGroups: {}
  plugins: []
  draft: function(part) {
    // same as it always was
  },
}
```
In this way, all the info about what a part needs and what it does is in one place and it's all added to the design at the same time. We don't have to add draft methods to the prototype of a Pattern because they're already attached to the part they pertain to. Parts can add any of their own pattern configuration data that the designer finds useful. This encourages a more modular approach to parts, and possibly opens the door for pack support. If you add a part from another design, it brings in the measurements, plugins, options, etc. that it needs so that you don't have to figure out what those are and bring them in separately.  It's still an option to put those things directly into the pattern configuration instead, but it's not preferred if you think someone might benefit from reusing parts of your design elsewhere.

Now your `index.js` looks something like this for a from-scratch design
``` javascript
// index.js
import Back from './back.js'
import {DesignConfig} from '@freesewing/core'
import SomePlugin from '@freesewing/plugin-someplugin'

// start a new config with the data you want and plugins you want
const config = new DesignConfig({
  name: 'myPattern',
  measurements: ['hips']
}, [SomePlugin])

// add parts directly to the config
config.addPart(Back)

// export as a design, but also export the config so it can be inherited
const myPattern = config.toDesign();
export {config, myPattern}
export default myPattern;
```

or it can look like this for a design that inherits from an existing design
``` javascript
// index.js
import {config as aaronConfig} from '@freesewing/aaron'
import Back from './back.js'
import SomePlugin from '@freesewing/plugin-someplugin'

// clone the config
const config = aaronConfig.clone()

// add new plugins to existing config
config.addPlugins(SomePlugin)

// hide a part from aaron
config.updatePart('front', {hidden: true})
// add another measurement
config.apply({measurements: ['hips']})
// add another part
config.addPart(Back, true);

// export as a design, but also export the config so it can be inherited
const myPattern = config.toDesign();
export {config, myPattern}
export default myPattern;
```

The core `DesignConfig` class looks like this:
``` javascript
class DesignConfig {
  config = {
    name: 'pattern',
    options: {}
    optionGroups: {}
    measurements: []
    optionalMeasurements: []
    version: '1.0.0',
    parts: {}
  }

  plugins = []
  // hold the raw part configs so that they can be overwritten easily when inheriting
  partConfigs = {}

  constructor(config = {}, plugins = []) {
    // merge any configuration that's given
    this.apply(config)
    // add any plugins
    this.addPlugins(plugins)
  }

  /** clone the config */
  clone() {
    const cloned = new DesignConfig(this.config, this.plugins);
    cloned.apply(this.partConfigs, cloned.partConfigs);
    return cloned;
  }

  /** object or array merging utility */
  apply(applyFrom, applyTo = this.config, replace=false) {
    // some logic to deep merge objects and concat arrays without duplicates
  }

  /** add plugins to the plugin array */
  addPlugins(...plugins) {
    this.apply(plugins, this.plugins)
  }

  /** 
   * add a part's configuration
   * if there's already a config for a part of that name, replace dictates whether it is merged or fully replaced 
   * */
  addPart(partConfig, replace=false) {
    if ((!this.partConfigs[partName] || replace) && !newConfig.draft) {
      console.warn(`trying to configure ${partConfig.name}, but there is no draft method for the part`)
      return
    }

    this.partConfigs[part.name] = this.partConfigs[part.name] || {}
    this.apply(this.partConfigs[part.name], part, replace)
  }

  /** utility for updating a part's configuration. calls addPart under the hood */
  updatePart(partName, newConfig) {
    if (!this.partConfigs[partName] && !newConfig.draft) {
      console.warn(`trying to update configuration for ${partName}, but there is no draft method for the part`)
      return
    }

    addPart({...newConfig, name: partName})
  }

  /** convert the config to a Design */
  toDesign() {
    // start with the config and plugins as they stand so far
    const compiledConfig = cloneDeep(this.config)
    let compiledPlugins = [...this.plugins]

    // go through the parts. we do this here instead of on add so that there are no lingering side effects from part configurations that were overwritten
    for (var part in this.partConfigs) {
      // the part can add its own data to any part of the configuration. this is to encourage a more modular approach to parts, and possibly allow for pack support. 
      let {measurements, optionalMeasurements, plugins, options, optionGroups, ...part} = partConfig

      if (measurements) this.apply(measurements, compiledConfig.measurements)
      if (optionalMeasurements) this.apply(optionalMeasurements, compiledConfig.optionalMeasurements)

      if (plugins) this.apply(plugins, compiledPlugins)

      if (options) this.apply(options, compiledConfig.options)
      if (optionGroups) this.apply(optionGroups, compiledConfig.optiongroups)

      // TODO maybe we double check that its dependencies and injections are also present before we add it?
      // add the part's data to parts
      compiledConfig.parts[part.name] = part
    }

    // create a design using the compiled config
    const design = new Design(compiledConfig)
    // add the plugins
    compiledPlugins.forEach((p) => design.use(p))
  }
}
```

**Comment by woutervdub** [comment]
Looks interesting.

Does this allow for multiple inheritance? Like give me the back from aaron, and the sleeves from carlton. (yes, there would obviously be issues with the sleeve fitting the opening, but I'm just exploring the principle)

Could this allow for creating 'interfaces' between parts? Like we define that a body block has/implements a 'neckOpening' and a 'sleeveOpening' interface, and a collar or sleeve block reacts to that by fitting itself to that interface.

**Reply by eriese** [reply]
Yes, I think it actually makes multiple inheritance easier, especially if we're all being responsible about including all the configuration needs with parts. You could do something like
``` javascript
import {config as carltonConfig} from '@freesewing/carlton'
import {config as aaronConfig} from '@freesewing/aaron'
import {DesignConfig} from '@freesewing/core'

const config = new DesignConfig({name: 'frankenShirt'})
config.addPart(carltonConfig.parts.sleeve)
config.addPart(aaronConfig.parts.back)
```

and then those parts make sure that the necessary measurements, settings, plugins, etc. that make them work are also added to the pattern

**Comment by woutervdub** [comment]
Does the inherited part get the info from the 'new' config?

Let's say I inherit a block that sets the 'chestEase' option to 0%. In my new design, I set that option to 10%. Will the block get the 10% value for the option?

If so, how do we deal with changing maximum and minimum values that could invalidate the inherited part's code?

**Reply by eriese** [reply]
the block will get the 10%. 

this is a good question that needs a better answer. In the proposal right now there's no enforcement of the order in which parts get added, which means that if multiple parts are adding the same option, whichever is added last is the one that will be used. 

Two ways of handling this that I can think of, maybe both are used:
1) Options brought by the parts are only added if the pattern doesn't already have an option of that name, so explicitly setting the option in the pattern config is essentially the designer manually resolving the conflict.
2) When we merge in an option with a min and a max, we set them to the min and max of offered values. I'm not sure what happens with the default value, but it's like this:
``` javascript
{pct: 10, min: 5, max: 50}.merge({pct: 15, min: 10, max: 60}) == {pct: ???, min: 5: max: 60}
```

**Comment by joostdecock** [comment]
I'm happy we're having this discussion. I'm on my phone now, but I'll weigh in on this tomorrow.

@woutervdub we refer to this the functionality you refer to as *packs*. It's on our roadmap, and it's what @eriese refers to when they say *possibly open the door to pack support*.

**Comment by joostdecock** [comment]
Before going into what I think are good ways to handle adding parts, I feel I should summarize the situation up until this point.

## Adding parts today

Parts are currently added to the `freesewing.Part` prototype. It is inspired by the way JS works (it's prototypical inheritance all the way down), but we don't really expose an API for this, instead just relying on the language features:

```js
Aaron.prototype.draftFront = (part) => draftFront(part)
```

When extending the prototype, there's a risk of collisions with existing properties on the object. So we worked around that by adopting a rule that for part `name` the method should be added to `draftName`.
Then core will just call `draftName()` for each pattern.

That works, but it's been bothering me for a while because for people who are new to Javascript, all that prototype boilerplate is hella confusing.

I'd like to simplify this and expose a clean(er) API to add parts, and hide some of the complexity in core.

## Run time vs build time attaching

This distinction between *run time* and *build time* also comes up in plugins. It's because you're basically dealing with different things depending on where you are in the life cycle of a pattern.

- **When developing a pattern** you create a new `Design` and add plugins (or parts) to it
- **When using a pattern** you add plugins (or parts) to the `Pattern` object, which is an instantiated `Design`

Technically, both are possible today. But adding parts is almost exclusively done at build time.
The alternative, adding parts at run time, is possible but requires a very good understanding of the core library. Case in point: The first time this was ever done is [when I wrote a plugin for it for the custom layout in our new frontend](https://github.com/freesewing/freesewing/blob/develop/sites/shared/components/workbench/layout/print/plugin.js#L18).

My point is: Adding parts at run time is currently not supported. It's merely possible for those who are not afraid to bend core to their will.

## What I want, and what I don't want

I would like to look into possibilities for making it easier/supported to add parts at run time.
And since we're doing that, I want to use the occasion to simplify how we add parts at build time too

That being said, there's a lot of things that I would want, and many of them would requires breaking changes to the core library.
Breaking changes are not really a problem as such, but we should try to group them and then put out a new major version.
I don't feel like now is the time for that, because we have a lot of work on our plate (migrate strapi, build new backend, build new frontend, convert our svg => pdf pipeline, migrate from monodb to sqlite, migrate all users, switch over our production website). and until that is dealt with, we're going to have to keep supporting the current/legacy site and our new work. Something which is already causing a lot of strain as illustrated by the time it takes for changes to make their way to freesewing.org.

So, first we have to get all this stuff done. Then we can think about version 3 and breaking changes.
To summarize:

- What I want is the ability to add parts both at build time and at run time in an elegant way
- What I don't want are breaking changes

## Why none of this is urgent

I feel this is something that gets overlooked. But the main reason this is coming up _now_ is that @eriese is working on the cutting layout. And they'd like to support things like duplicating parts and so on, which means we need an elegant way to add parts at run time.

However, while I appreciate all the work and their investment in this feature, I do not consider it crucial.
I wanted our new site to support custom (print) layouts by the users. That is a very often requested feature, and something that I really wanted to include. So that's why I started working on it.
When I started on it, I added a link in the side menu to also foresee a cutting layout since if we can do custom layouts, why not also support one for cutting, right? But after that, I didn't really look into it since my focus was on the print layout.

Now @eriese is working on the cutting layout and it this that is driving this request for changes to the core library.
But, and I hope this is not controversial, cutting layout is *nice to have* but it's not a blocking factor for me.
I'd like to get the new site out the door as soon as possible (which won't be soon no matter what).
Because once that's done, we hopefully will have an easier job maintaining things and can start thinking about all the stuff we want to see in v3.

So while I am willing to facilitate improving the API to make run time addition of parts possible, I don't think now is the time to introduce breaking changes that would only widen the rift between what our users get to play with, and what we are working on.

## My proposal

Taking all this into account, I would prefer an approach that simplifies the development experience for our designers, remains backwards compatible, yet somehow makes adding parts at run time possible.

It would look something like this:

**index.js**
```js
import freesewing from 'freesewing/core'
import plugins from '@freesewing/plugin-bundle'
import bustPlugin from '@freesewing/plugin-bust' // Note: conditional plugin
import { version } from '../package.json'
// This should become a named export in v3
import configHelpers from '@freesewing/config-helpers'
const { pctBasedOn } = configHelpers

// Parts should be named exports so that we can simple import one part from another pattern
import { base } from '@freesewing/brian'
import back from './back'
import front from './front'

// Create design with one big initializer object
const Aaron = new freesewing.Design({
  plugins: plugins,
  conditionalPlugins: [ { 
    plugin: bustPlugin, 
    /* Check to see whether we should load the bust plugin
     * Only of the `draftForHighBust` options is set
     * AND the highBust measurement is available
     */
    condition: (settings = false) =>
      settings &&
      settings.options &&
      settings.options.draftForHighBust &&
      settings.measurements.highBust
        ? true
        : false
  } ],
  parts: { base, back, front },
  version,
  name: 'aaron',
  design: 'Joost De Cock',
  code: 'Joost De Cock',
  department: 'tops',
  type: 'pattern',
  difficulty: 2,
  optionGroups: {
    fit: ['chestEase', 'hipsEase', 'stretchFactor', 'draftForHighBust'],
    style: [
      'armholeDrop',
      'backlineBend',
      'necklineBend',
      'necklineDrop',
      'shoulderStrapWidth',
      'shoulderStrapPlacement',
      'lengthBonus',
    ],
  },
  measurements: [
    'biceps',
    'chest',
    'hpsToWaistBack',
    'waistToHips',
    'neck',
    'shoulderSlope',
    'shoulderToShoulder',
    'hips',
  ],
  optionalMeasurements: ['highBust'],
  options: {
    // Constants
    brianFitCollar: false,
    collarFactor: 4.8,
    acrossBackFactor: 0.97,
    backNeckCutout: 0.05,
    bicepsEase: 0.05,
    shoulderEase: 0,
    collarEase: 0,
    frontArmholeDeeper: 0,
    armholeDepthFactor: 0.6,
    shoulderSlopeReduction: 0,

    // Percentages
    armholeDrop: { pct: 10, min: 0, max: 75 },
    backlineBend: { pct: 50, min: 25, max: 100 },
    chestEase: { pct: 8, min: 0, max: 20, ...pctBasedOn('chest') },
    hipsEase: { pct: 8, min: 0, max: 20 },
    lengthBonus: { pct: 10, min: -20, max: 60 },
    necklineBend: { pct: 100, min: 40, max: 100 },
    necklineDrop: { pct: 20, min: 10, max: 35 },
    stretchFactor: { pct: 5, min: 0, max: 15 },
    shoulderStrapWidth: { pct: 15, min: 10, max: 40 },
    shoulderStrapPlacement: { pct: 40, min: 20, max: 80 },

    // draft for high bust
    draftForHighBust: { bool: false },
  },
})

// Named exports
export { 
  config: Aaron.config,
  info: Aaron.info,
  Aaron, 
  front, 
  back 
}

// Default export
export default Aaron
```
**front.js**
```js
/*
 * Parts can simple export a method (like today) 
 * or an object where `draft` holds the method 
 * In this case, other config options that can be set are
 *   from: The new inject
 *   after: dependencies
 *   hide: true or false
 */
export default {
  from: 'base', 
  draft: function (part) {
    // Part code goes here
  }
}
```
**Frontend**
```js
import Aaron from '@freesewing/aaron'

const draft = new Aaron(gist).addPart({
  from: 'front',
  draft: function(part) {
    // ...
  }
})
```

## Notes
A few things that are worth pointing out:

**Note that I did not import a config**
If you feel this is a lot of code, please keep in mind that rather than importing `config.js` I've included it all in `index.js`, and I wanted to also show things like conditional plugins. In other words, don't make the mistake thinking this is more verbose, it is not.

**Design contructor takes one parameter**
The `Design` constructor takes one parameter that holds, well everything. Today, it takes three arguments:
 - config
 - plugins
 - conditionalPlugins

Instead, we'll just squash them all together in one object. All config top-level properties become top-level properties in the object.
`plugins`, `conditionalPlugins` hold what you'd expect. `parts` now holds an object where the keys are the names of the parts and the values hold one of:
- A method that drafts the part
- An object where the `draft` property holds the method that drafts the part, and the `from`, `after`, and `hide` properties configure `inject`, `dependencies` and `hide` in the config

**Config at the part level**
That's right, while we're changing things, I would propose to give new names to:

- `inject`: This is now set through `from`. That name does not make sense in a pattern-wide config file, but does make sense here since this is now set at the part level. This part `from` another part seems intuitive to me
- `dependencies`: This is now set through `after`. As in, handle this part `after` this/these parts have been done.

> Note that unlike the proposal above, these are the only things that should be configured from the part. Although I guess we can fold other keys passed here into the config

**Split between `config` and `info`**
The Design constructor will sort whatever you pass to it into two buckets:

- `config`: Holds everything required for the pattern to work
- `info`: Holds everything else. Currently this is things like `name`, `version`, `difficulty`, `optionGroups` and so on. None of this is needed for the pattern to work. It's things that we use for our frontend integration. But other people might want to add other things and this would allow you to add whatever.

**Parts should be named exports**
We should export all parts as named exports. This means we can do this awesome thing:
```js
import  { base } from '@freesewing/brian`
```

That does mean we need to avoid name clashed, so I guess this would make naming your part either `config` or `info` unsupported.
I don't think that's a big problem.

**Add a `pattern.addPart()` method**
It should take input in line with what parts export. So either a function, or an object with config+function in `draft`.

**Limitations**
- You can't name your part `config` or `info`
- This would support adding parts both at build time and run time. But you need to add all build time parts in the constructor. You can't do something like `Aaron.addPart()`.

## Why not just accept the proposal above?

- I don't like using classes. Javascript doesn't really have classes anyway, just some sort of emulation
- I prefer polymorphism to handle backwards compatibility, rather than pushing a bunch of new objects/classes into core
- Spreading your config like options and measurements and so on across a variety of parts seems like it would add a lot of mental overhead
- Adding parts and plugins to something called `DesignConfig` (or calling `config.addPart(Back)` seems counterintuitive to me. Is it config or not?
- This `const myPattern = config.toDesign();` mixes up terminology. We should never use `pattern` here, we're working with `design` so I guess it should be `const myDesign = config.toDesign()` which ... I don't like
- I feel this replaces a bunch of boilerplate with a bunch of different boilerplate. So I much prefer the single call to `new Design()`. You create a design, and then you export it. Done.

**Reply by eriese** [reply]
As I said on the call, I'm into this proposal. It covers everything that was really important to me and keeps them more in the style of code that people will be accustomed to if they've already developed for freesewing. I think there's room for packs to be plugins using this route.

I still think that it'd be nice to have a mechanism to simplify the configuration aspect of part inheritance (i.e. measurements, options, etc). I agree that mine is a lot of mental overhead, but I'm interested in whether we can come up with something.

My main goal with using a class for what I think would be better called `DesignConfigBuilder` was to make it easier for people to override inherited config stuff instead of using a lot of object destructuring. This could be replaced by an importable utility method that does what `apply` would do within `DesignConfigBuilder`

**Comment by eriese** [comment]
What if the part config just had an array of measurements used by the part and an array of the names of options used by the part? that way you're not having to duplicate option configuration in multiple parts, but we could have a method called, like, `extractPartForInheritance(design, partName)` that uses that data to go in and get everything that needs to come with the part?

Either that or `extractPartForInheritance` runs the draft method with a dummy part that has proxies that record when measurements are asked for and when options are asked for. The tricky part with that one is how you get it to visit all possible branches of the draft function

**Comment by joostdecock** [comment]
The more I think about this, the more I feel we might want to land on some sort of a two-tier system.

We've made some good progress on adding parts. The issue that we're now trying to solve is making this *just work*:

```js
import { sleeve } from '@freesewing/brian'
```

The `sleeve` part depends on the `sleevecap` part. So to make this *just work* we need more than just adding all the options and measurements required by the `sleeve` part to the (named) export. We'd also somehow have to auto-add all the parts it depends and and their configuration.

I'm not saying that's impossible -- I have some embryonic ideas -- but I do feel that we risk making pattern design a lot more complicated if we're going to expect designers to do this.

Instead, I would suggest that in our tutorial for new designers we don't; include this and just switch to a new way of adding parts where only the inject/hide/dependencies are kept along the part code.

Then, we can take that design and untangle the config. This can be done by more experienced developers, and it allows us to iterate faster on tools to help accomplish that, like running the design in some sort of introspection mode that proxies all config access and tries to auto-generate a segregated configuration.

In other words, I'm in favor of ending up there, but I feel we should not demand from designers that they reach that point on their own.

As for config segregation, we obviously don't want to repeat option settings (for example) in 2 places.
Having a list/array of options works around that, but doesn't magically include that config with the part.

But it should be doable to write a script/plugin that unwraps the config into atomic items. For example:

`config/options.js`
```js
export const chestEase = { pct: 8, min: 0, max: 16 }
export const waistEase = { pct: 10, min: 0, max: 18 }
```

Then in the part (and this should also he possible to auto-generate) you have:

```js
import {
  chestEase,
  waistEase
} from '../config/options.js'

export default {
  options: { chestEase, waistEase },
  draft: function (part) {
    // Original design code here
  }
}
```

It's essentially just listing the options, but instead of an array of option names, you have the actual config in the part. You can also reuse it in other parts, yet it's still managed in one central place.

I suspect that we will improve our tooling to make nore and more of this automatic, and I think that's exciting. But I would caution to not push this complexity on aspiring designers.

Anyway, I'm happy to look into this and see what I can come up with, as auto-generating this is essentially independent of how parts are added.

**Reply by eriese** [reply]
> Then, we can take that design and untangle the config. This can be done by more experienced developers, and it allows us to iterate faster on tools to help accomplish that, like running the design in some sort of introspection mode that proxies all config access and tries to auto-generate a segregated configuration.
> 
> In other words, I'm in favor of ending up there, but I feel we should not demand from designers that they reach that point on their own.

Yes, agreed that the goal is to keep what designers have to specify as low as possible, and the number of places that they have to specify it as close to 1 as possible.

> As for config segregation, we obviously don't want to repeat option settings (for example) in 2 places. Having a list/array of options works around that, but doesn't magically include that config with the part.

My thinking was that for the time-before-introspection-mode, the list/array of options tells the `extractForInheritance` function what to extract and then it can go and get it from the config. Instead of true magic, we have the magic of a utility we've given some instruction to. Because of course I'd personally centralize my options and import them to the parts that need them, but I don't expect that approach to be intuitive to everyone. The reason what I propose feels simpler to me is that you don't have to remember to import and export stuff. Just, as you're going through writing your draft code, you're like, "Oh, I used this measurement/option, better add it to the list"

So what I'm suggesting is
``` javascript
// brian/src/sleeve.js
export default {
  from: 'sleeveCap'
  options: ['thisOption', 'thatOption']
  measurements: ['measie1', 'measie2'],
  draft: function(part) {...}
}
```

``` javascript
// brian/src/sleeveCap.js
export default {
  options: ['optionTheThird'],
  measurements: ['measie3'],
  draft: function(part) {...}
}
```

``` javascript
import {config as brianConfig} from '@freesewing/brian'
import {extractForInheritance, mergeConfigs} from '@freesewing/core'
let config = {}
const brianSleeve = extractForInheritance(brianConfig, 'sleeve'}
/** 
  {
    parts: {
      sleeve: {
        from: 'sleeveCap'
        options: ['thisOption', 'thatOption']
        measurements: ['measie1', 'measie2'],
        draft: function(part) {...}
      },
      sleeveCap: {
        options: ['optionTheThird'],
        measurements: ['measie3'],
        draft: function(part) {...}
      }
    },
    options: {
      thisOption: {...whatever it was from brian's config},
      thatOption: {...}
      optionTheThird: {...}
   },
  measurements: ['measie1', 'measie2', 'measie3']
  }
*/
config = mergeConfigs(config, brianSleeve)
```

**Comment by joostdecock** [comment]
Note that I have started kicking the tires of these ideas in #2546 

I am aware that there is still a bit of convergence left to be done between myself and @eriese -- or rather between our respective proposals -- and I do not want to stifle that debate by firing up my code editor. In other words, just because I started exploring my ideas in code doesn't mean this discussion is closed.

However, I find there is only so much I can theorize about what would work best. At the end of the day, the best test is trying to make it work, so that's what I've started on. 

I suspect this will take a while, so I've put up a PR (#2546) for those who are interested. I will try to keep it documented with the more relevant changes.

**Comment by anna-puk** [comment]
Some of this discussion went over my head, but I'll weigh in anyway. :)

Things I liked:
- the idea of 'less boilerplate code' - I tend to copy/paste it and hope for the best, but the more code I ignore, the less confident I feel
- getting rid of (if I understood this correctly) config.js and putting everything in index.js - less files = less places to consider when looking for / figuring out where to put things
- the ability to re-use or even mix-and-match parts: I don't quite understand what will and won't be easy to do in the new version, but for designers, it would be awesome to know that e.g. a bell sleeve I might design for Diana could easily be reused somewhere else.

Things I'm not so sure about:
- defining options and measurements per part: how does this work when using 'inject'? For Unice, the gusset width is determined while drafting the 'front' part, and passed via 'store'. Does that mean the 'gusset' part would require no measurements at all? What if I use the same option for multiple parts?
- having imports 'just work' multiple levels deep, e.g. the sleevecap being implicit in the import of sleeve - I tend to lose track when dependencies get nested like this. (I know this kind of thing is common and well-regarded, though, so I guess this is a minority opinion.) If I'm modifying an existing design, I can probably point out in the svg which points I want to mess with. To find the relevant code, I'd first look at the [partname].js file. If I don't find it there, I need to check both preceding parts and any 'inject's. With additional levels, the number of files increases even more... Perhaps this is just the wrong approach?

Thing I'd like to see changed:
- the 'from' and 'after' names - while easy enough to remember, these are not intuitive to me and if I came across them without knowing these concepts, I'd have no clue what this does. Can we elaborate a bit, maybe 'derivedFrom' and 'draftedAfter'?

I also want to make a plea to consider not just how to make it easier to _write_ code, but especially how to make it easier to _read_ it. (As someone pointed out to me once, you write code only once, but multiple people can read it multiple times.) I tend to benefit from this myself (after three months, I barely remember what I was doing), but it also benefits other developers who are trying to learn from existing code, and makes it easier for someone to take over maintenance. For me, this would be a reason not to do too many things 'under the hood'. If people don't know that it's happening, they won't look it up in the documentation - they couldn't, for to them it has no name.

**Reply by joostdecock** [reply]
**How does it work to define options & measurements per part**

> For Unice, the gusset width is determined while drafting the 'front' part, and passed via 'store'. Does that mean the 'gusset' part would require no measurements at all?

- You would define the `front` part with the measurements/options it requires
- You would then define the `gusset` part to be dependent on the `back` part
- When people want to (re)use your `gusset` part, Core should follow the dependency chain and pull in all required measurements/options from  `back` too

So in short: You would not need to define any measurements on the gusset part in this case
 
> What if I use the same option for multiple parts?

My current approach is to define them as a named export centrally, and import them in each part so you can include them everywhere but still manage them centrally.

**having imports 'just work' multiple levels deep**

Core will follow the dependency chain, so if you pull in one part that relies on another, it will pull in that one too.


**Can we change `from` and `after` to `derivedFrom` and `draftedAftert`**

The names are not set in stone. But I'd argue that `from` is a marked improvement over the current `inject` and `after` is more intuitive than the current way dependencies are impemented:

```js
dependencies: {
  front: "back",
}
```
☝️ does `front` depend on `back`, or is it the other way around? Do you know without checking the documentation?

I feel there's a balance to be struck between being terse and being self-explanatory. And I'm not convinced `derivedFrom` is such a marked improvement over `from` that it justifies having to juggle cases and force everyone to type the word `derived` which I suspect is not the most obvious word to people for whom English is a barrier.

That being said, I am happy to consider `from` and `after` to be placeholders for now, and we can continue this debate.

**I also want to make a plea to consider not just how to make it easier to write code, but especially how to make it easier to read it**

> this would be a reason not to do too many things 'under the hood'

I find this hard to reconcile with your earlier sentiment that you are happy to see the boilerplate code go away, since you just copy/paste it without understanding.

That boilerplate *does something*. So if we want to get rid of it, we're going to have to do that *something* under the hood.

- No boilerplate of things that are hard to understand
- Don't hide complexity behind the curtains

Pick one. You can't have your cake and eat it. 

My guiding principle in this is that I want to make it easy for people without prior programming experience.
In practical terms, that means that it's ok to expect people to familiarize themselves with the FreeSewing (core) API.
It's not ok to expect people to be familiar with things like prototype inheritance, because that is not a FreeSewing thing. That is a Javascript thing.

A good illustration of this difference is your excitement that there is no longer a `config` file and an `index` file.
This is an arbitrary distinction. You always had the possibility to put everything in 1 file. Because these are all features of Javascript (the language) rather than FreeSewing-specific.

That being said. Aaron right now:
```js
/* Check to see whether we should load the bust plugin
 * Only of the `draftForHighBust` options is set
 * AND the highBust measurement is available
 */
const condition = (settings = false) =>
  settings &&
  settings.options &&
  settings.options.draftForHighBust &&
  settings.measurements.highBust
    ? true
    : false

// Create design
const Aaron = new freesewing.Design(config, plugins, { plugin, condition })

// Attach draft methods to prototype
Aaron.prototype.draftBase = function (part) {
  // Getting the base part from Brian
  return new Brian(this.settings).draftBase(part)
}
Aaron.prototype.draftFront = (part) => draftFront(part)
Aaron.prototype.draftBack = (part) => draftBack(part)
```

Aaron in my feature branch:
```js
const Aaron = new freesewing.Design({
  ...info,
  measurements,
  optionalMeasurements,
  options: { ...options },
  parts: { back, front },
  plugins: pluginBundle,
  conditionalPlugins: {
    plugin: bustPlugin,
    condition: (settings=false) =>
      settings?.options?.draftForHighBust &&
      settings?.measurements?.highBust
      ? true : false
  }
})
```

Which one do your find easiest to read?

**Comment by BenJamesBen** [comment]
It worries me that we're discussing code implementation before the requirements have been pinned down. How do we want it to work in the end? Once we know that, we might have a better understanding of how it might be best implemented in code.

I added some thoughts about how things should work in the "Discuss: A modular approach to patterns" discussion: https://github.com/freesewing/freesewing/discussions/1107#discussioncomment-3386175 .

**Reply by joostdecock** [reply]
I should start by saying that I think [your comments](https://github.com/freesewing/freesewing/discussions/1107#discussioncomment-3386175) are excellent. But they read a bit like a *greenfield implementation what if* type of situation. By which I mean: The scope is wider, and everything is put on the table to see what could be accomplished.

I think that's a fun/exciting discussion to have, and I certainly don't want to shut it down. But (you knew there was a but coming, didn't you?) I don't think that's what we're trying to do here.

We have [a roadmap](https://github.com/freesewing/freesewing/discussions/1278) which outlines where we want to go. It includes such things as **Support for packs** which includes:

> Good as that sounds on paper, it's no so clear how this would work under the hood.
On the one hand, there's the mechanism of extending patterns, which requires a bunch of code and attaching methods to prototypes and whatnot, which is scary for people new to Javascript. You also have to make sure you're taking over all the relevant configuration in your own designs' config and whatnot.
>
> On the other side of the spectrum are plugins. You can just npm install them and load them with a .use() call. Handy.
>
> I'd love to look into finding a way that you can load a pack more like a plugin, and less like a pattern. In other words, make it easier to extend patterns with this sort of thing.

Now you could argue that these requirements are rather vague, but I don't think there's any confusion of where we want to go. We want to make it as simple to attach a part to a pattern than it is to use a plugin.

This discussion is not only about that, it is also about how we can do that **without causing breaking changes**.
Which somewhat limits the scope because it's not a green-field implementation, it's a study of *how could this work* while ensuring backwards compatibility.

Apart from that, I think you should see the fact that code is being written more as my own limits of how much I can think about this in abstract terms. Like, I want to kick the tires, see how it feels, what works and what doesn't.
So writing code is just my way of thinking about how this could work.

I typically start by writing what I would like it to look like. In this case:

```js
import  { sleeve } from '@freesewing/brian'

const part = {
  from: sleeve,
  draft: part => {
     // Do stuff here
  }
}
```

⬆️ This satisfies the requirements outlined in **Support for packs** so how do we get it to work with FreeSewing core?

I think all of your comments are interesting and worthwhile, and I would really don't want to discourage these fresh ideas you bring to the table because they are really valuable.

But unlike the stuff above, they are not on our roadmap. So I guess the point I'm trying to make is that:

- Yes, code is being written, but that's just because I think *talk is cheap*
- It should not be a surprise or cause for concern that we are working on things that are on our roadmap
- While I agree the description on the roadmap is not very detailed, I do feel like we know where we want to land with this

Perhaps I attach to much importance to your use of the word **worry** here, but it makes me feel like we've done something wrong.
And while I want to encourage everybody to speak out when they feel something is not right -- because we all make mistakes and should be appreciative when they are brought to our attention -- I don't think I understand what has you worried here.

This won't be the last change we make. There will be room for your ideas in the future. 
I don't think we are closing any doors here that would hinder adoption of any of these ideas in the future.

So I hope your worries aren't too profound, and I would encourage your to flesh out your ideas and get them onto our roadmap.

**Comment by joostdecock** [comment]
Reminder that #2546 is the best place to check the current status of my work on this

**Comment by BenJamesBen** [comment]
Thanks. I think that this project is just using a much more agile/iterative process than I am used to.

**Comment by joostdecock** [comment]
Stale. Closing. But I guess we have settled on how to handle this in v3 so it's all good.
